# Metagenomic pipeline for Bat samples
This repository provides data and scripts used for the [MRA article]()
The workflow contains scripts written in C++ and C#, the user may need to adjust the file path manually, to do so, the [Microsoft VisualStudio](https://visualstudio.microsoft.com/zh-hans/) or other compatible IDEs are needed.

## **Raw sequencing data QC**
The Sequencing data was generated by the DNBSEQ-T7 V3.0 platform and contains 99.89 million paired-end reads of 150 nt [SRR31145840](https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR31145840&display=metadata). 

Raw reads were filtered using [fastp](https://github.com/OpenGene/fastp) to remove adapters, polyX sequences, low-quality tails, and reads shorter than 50bp. Using the following command:

>_fastp -i $Fq_R1 -I $Fq_R2 -o p1.fq -O p2.fq -l 50 -x -w 20 --detect_adapter_for_pe --cut_tail --cut_tail_mean_quality 20 > $Result.qc 2>&1_

## **Host species determination**

To determine the Bat species, the reads after QC were mapped to a eukaryotic mitochondrial genome database (a fasta file), provided by [Crits-Christoph et al. Cell 2024](https://zenodo.org/records/12571327) using [Minimap2](https://github.com/lh3/minimap2).

>_minimap2 -t 4 -a -x sr Database/mitochondria_dereplicated93_final.fasta p1.fq p2.fq | samtools view -@ 4 -F 4 > mtDNA.sam_

Then, the Microsoft Excel was used to calculate the distribution of the third column (Accession number of mapped reference) in the .sam file.

## **Host genome depletion**

As no complete genome was available for _Pipistrellus abramus_, we chose all four reference genomes from the same genus ([GCA_903992545.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_903992545.1/), [GCA_949987585.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_949987585.1/), [GCA_963693515.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_963693515.1/), and [GCA_014108245.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_014108245.1/)) and used them for host genome depletion with [Bowtie2](https://github.com/BenLangmead/bowtie2).
>_cat GCA_903992545.1_genomic.fna GCA_949987585.1_genomic.fna GCA_963693515.1_genomic.fna GCA_014108245.1_genomic.fna > Pipistrellus.fna_

>_bowtie2-build Pipistrellus.fna Pipistrellus_

>_bowtie2 -p 20 -x Pipistrellus -1 p1.fq -2 p2.fq | samtools view -@ 20 -b -f 12 -F 2 | samtools sort -@ 20 -o Dehg38.bam_

>_samtools fastq -@ 20 -1 Dehost.p1.fq -2 Dehost.p2.fq Dehg38.bam_


## **Taxonomy classification**

The clean reads were classified using [Kraken2](https://github.com/DerrickWood/kraken2) with its standard Refseq database.
>_kraken2-build --standard --threads 24 --db KrakenDBStandardMini_

>_kraken2 --db KrakenDBStandardMini --threads 2 --report Dehost.krakenreport --paired Dehost.p1.fq Dehost.p2.fq > KrakenReads.txt_

## **De novo assemblying**

De novo assemblies were generated using [MEGAHIT](https://github.com/voutcn/megahit).

>megahit -1 Dehost.p1.fq -2 Dehost.p2.fq -m 0.7 -t 20 -o megahit_out

## **Mapping contigs to reference**
Contigs longer than 500bp were retained and amplified based on their average depth using **C# script AmplifyMegahitcontig** before mapping to the _Staphylococcus nepalensis_ reference genome [GCA_002442935.1](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_002442935.1/) using Minimap2.

>_minimap2 -t 4 -a GCF_002442935.1_ASM244293v1_genomic.fna final.contigs.fa | samtools view -@ 4 -b -F 4 -h | samtools sort -@ 4 -o contigMapRef.bam_

## **Mapping reads to reference**

The reads assigned under genus _Staphylococcus_ were extracted using [SeqTK](https://github.com/lh3/seqtk) and mapped to the same reference with Minimap2.
>_python3 Script/filter.py KrakenReads.txt --taxid 1279--taxid_col 3 | cut -f 2 > 1279.reads.id_

>_seqtk subseq Dehost.p1.fq 1279.reads.id > 1279.R1.fq_
>_seqtk subseq Dehost.p2.fq 1279.reads.id > 1279.R2.fq_

>_minimap2 -t 4 -a -x sr GCF_002442935.1_ASM244293v1_genomic.fna 1279.R1.fq 1279.R2.fq | samtools view -@ 8 -b -F 4 -h | samtools sort -@ 8 -o readsMapRef.bam_

## **Merge mapping result**

We applied a mixed reference-based genome assembly tec, which uses de novo assembled contigs as long reads, as the backbone. And use short reads to fill the gaps.

First, convert bam file to readcounts file using [Varscan2](https://github.com/Jeltje/varscan2).

>_samtools mpileup -a --reference GCF_002442935.1_ASM244293v1_genomic.fna -d 999999 -Q 20 -q 20 contigMapRef.bam > contigMapRef.mpileup_
>_samtools mpileup -a --reference GCF_002442935.1_ASM244293v1_genomic.fna -d 999999 -Q 20 -q 20 readsMapRef.bam > readsMapRef.mpileup_

>_varscan readcounts contigMapRef.mpileup --min-coverage 0 --output-file contigMapRef.readcounts_
>_varscan readcounts readsMapRef.mpileup --min-coverage 0 --output-file readsMapRef.readcounts_

Second, using the **C# script RefDenovoMixAssembly** to generate a combined readcount file.
>_Merged.readcounts_

## **Generate consensus genome**

The final consensus genome Bat2024WH was generated using **C++ script CallConsensusSequence**, with a minimum depth requirement of 5 and an allele frequency requirement of 50%.

>_g++ Script/CallConsensusSequence/CallConsensusSequence/CallConsensusSequence.cpp > CallConsensusSequence.out_

>_CallConsensusSequence.out -i Merged.readcounts -t 0.5 -m 5_